pkg/controllers/vdb/replicatedupgrade_reconciler.go

```diff
@@ -929,3 +968,176 @@ func (r *ReplicatedUpgradeReconciler) getSandboxPodFacts(ctx context.Context, do
+}
+
+// renameReplicaGroupBFromVdb will rename the subclusters in promoted-sandbox/new-main-cluster to
+// match the ones in original main cluster
+func (r *ReplicatedUpgradeReconciler) renameReplicaGroupBFromVdb(ctx context.Context) (ctrl.Result, error) {
+	// if replica group A still exists or subclusters in replica group B have been renamed, we skip
+	// renaming the subclusters in replica group B
+	if r.countSubclustersForReplicaGroup(vmeta.ReplicaGroupAValue) > 0 ||
+		r.areGroupBSubclustersRenamed() {
+		return ctrl.Result{}, nil
+	}
+
+	initiator, found := r.PFacts[vapi.MainCluster].FindFirstPrimaryUpPodIP()
+	if !found {
+		r.Log.Info("Requeue because there are no primary UP nodes in main cluster to execute rename-subcluster operation")
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	scNames := r.getSubclustersForReplicaGroup(vmeta.ReplicaGroupBValue)
+	for _, scName := range scNames {
+		newScName, found := r.scNameMap[scName]
+		// ignore the subcluster that has been renamed
+		if !found {
+			continue
+		}
+		// rename the subcluster in vertica
+		err := r.renameSubcluster(ctx, initiator, scName, newScName)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		// rename the subcluster in vdb
+		err = r.updateSubclusterNamesInVdb(ctx, scName, newScName)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+	return ctrl.Result{}, nil
+}
+
+// renameSubcluster will call vclusterOps to rename a subcluster in main cluster
+func (r *ReplicatedUpgradeReconciler) renameSubcluster(ctx context.Context, initiator, scName, newScName string) error {
+	opts := []renamesc.Option{
+		renamesc.WithInitiator(initiator),
+		renamesc.WithSubcluster(scName),
+		renamesc.WithNewSubclusterName(newScName),
+	}
+	r.VRec.Eventf(r.VDB, corev1.EventTypeNormal, events.SandboxSubclusterStart,
+		"Starting rename subcluster %q to %q", scName, newScName)
+	err := r.Dispatcher.RenameSubcluster(ctx, opts...)
+	if err != nil {
+		r.VRec.Eventf(r.VDB, corev1.EventTypeWarning, events.SandboxSubclusterFailed,
+			"Failed to rename subcluster %q to %q", scName, newScName)
+		return err
+	}
+	r.VRec.Eventf(r.VDB, corev1.EventTypeNormal, events.SandboxSubclusterSucceeded,
+		"Successfully rename subcluster %q to %q", scName, newScName)
+
+	return nil
+}
+
+// updateSubclusterNamesInVdb will update the names of subclusters in VerticaDB
+func (r *ReplicatedUpgradeReconciler) updateSubclusterNamesInVdb(ctx context.Context, scName, newScName string) error {
+	r.Log.Info("starting renaming subcluster in VerticaDB", "subcluster", scName, "new subcluster name", newScName)
+
+	updateSubclustersInVdb := func() (bool, error) {
+		// rename subcluster
+		for i := range r.VDB.Spec.Subclusters {
+			if r.VDB.Spec.Subclusters[i].Name == scName {
+				r.VDB.Spec.Subclusters[i].Name = newScName
+				return true, nil
+			}
+		}
+		return false, nil
+	}
+
+	updated, err := vk8s.UpdateVDBWithRetry(ctx, r.VRec, r.VDB, updateSubclustersInVdb)
```

Comment:
Not related to this but in sandbox_reconciler vdb is updated before the sandbox api call. Can you change that?